src/resmom/linux/alps.c

```diff
@@ -5817,15 +5817,61 @@ alps_cancel_reservation(job *pjob)
 {
 	char buf[1024];
 	basil_response_t *brp;
+	char	filename[MAXPATHLEN];
 
 	if (!pjob) {
 		log_event(PBSEVENT_SYSTEM, PBS_EVENTCLASS_NODE, LOG_NOTICE, __func__,
 			"Cannot cancel ALPS reservation, invalid job.");
 		return (-1);
 	}
+
+	/*
+	 * Get the name of the file that contains ALPS reservation ID
+	 * information.  We may need to read the file, or we just need to
+	 * unlink it.  Either way, we need the filename.
+	 */
+	snprintf(filename, MAXPATHLEN, "%s/aux/%s.alpsresv", pbs_conf.pbs_home_path, pjob->ji_qs.ji_jobid);
+	
+	if (pjob->ji_extended.ji_ext.ji_reservation < 0) {
+		/* 
+			* The job structure doesn't have ALPS reservation ID information.
+			* This can happen when we hit various race conditions.  To help
+			* with this issue, we have previously saved the ALPS resv ID
+			* in a file.
+			*
+			* The format of the file will consist of one line entry that
+			* has the string alps_resv_id a space and the ALPS reservation ID
+			* number.  We retrieve it here so the ALPS reservation can
+			* be canceled.
+			*/
+		FILE    *fp = NULL;
+		long    value;
+
+		fp = fopen(filename, "r");
+		if (fp == NULL) {
+			snprintf(log_buffer, sizeof(log_buffer), "Can't open the file %s", filename);
+			log_event(PBSEVENT_ERROR, PBS_EVENTCLASS_JOB, LOG_DEBUG,
+				(char *)__func__, log_buffer);
+			return (-1);
+		}
+		if (fscanf(fp, "alps_resv_id %ld", &value) != 1) {
+			log_event(PBSEVENT_ERROR, PBS_EVENTCLASS_JOB, LOG_DEBUG, 
+				(char *)__func__, "failed to read the alps_resv_id value");
+			(void)fclose(fp);
+			(void)unlink(filename);
+			return (-1);
+		}
+		(void)fclose(fp);
+		pjob->ji_extended.ji_ext.ji_reservation = value;
+	}
+
+	/* At this point we have no further use for the file that contains
+		* the ALPS reservation ID.  Get rid of the file.
+		*/
+	(void)unlink(filename);
+	
 	/* Return success if no reservation present. */
-	if (pjob->ji_extended.ji_ext.ji_reservation < 0 ||
-		pjob->ji_extended.ji_ext.ji_pagg == 0) {
+	if (pjob->ji_extended.ji_ext.ji_reservation <= 0 ) {
```

Comment:
Do we need this change?

Reply:
The case where pagg ID is zero, I saw it in when interactive job was terminated using ctrl +C. But we have already handled this condition and taken care of the orphaned reservation. So now we just have to check if there are any other reservation that needs to be canceled during job exit or clean up. So we don't check pagg ID.

Reply:
The case where pagg ID is zero, I saw it in when interactive job was terminated using ctrl +C. But we have already handled this condition and taken care of the orphaned reservation. So now we just have to check if there are any other reservation that needs to be canceled during job exit or clean up. So we don't check pagg ID.

Reply:
can there be other instances apart from an interactive job that the PAGG id should be checked for?

Reply:
I don't think there will  be any other case where PAGG ID is zero than than this.